akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -247,3 +224,78 @@ private[akka] final class HeadOptionStage[T] extends SinkStage[T, Future[Option[
     }, p.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] class QueueSink[T]() extends SinkStage[T, SinkQueue[T]]("queueSink") {
+  trait RequestElementCallback {
+    def requestElement(p: Promise[Option[T]]): Unit
+  }
+
+  type Received[E] = Try[Option[E]]
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+  val requestInProgress = new AtomicBoolean()
+  val buffer = new LinkedBlockingQueue[Received[T]](maxBuffer + 1)
+  val logicInitialized = new CountDownLatch(1)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val stageLogic = new GraphStageLogic(shape) with RequestElementCallback {
+      override def keepGoingAfterAllPortsClosed = true
+      override def requestElement(p: Promise[Option[T]]): Unit = {
+        implicit val ex = interpreter.materializer.executionContext
+        Future(buffer.take()).onComplete {
+          case Success(opt) ⇒ callback.invoke((opt, p))
+          case Failure(t)   ⇒ p.failure(t) //timeout
+        }
+      }
```

Comment:
Instead of this mix of different executions within the GraphStageLogic I’d propose to separate things out:

* have an external `val getter = Promise[(Promise[T] => Unit)]` that is fulfilled with an AsyncCallback from `preStart`
* `queue.pull()` then creates a `Promise[T]`, puts it into `getter.foreach(_(p))(sameThreadExecutionContext)` and returns the associated Future